# 4. Pre-Processing and Training Data Development

The Data Science Method

1.Problem Identification

2.Data Wrangling

3.Exploratory Data Analysis

4.Pre-processing and Training Data Development

    Create dummy or indicator features for categorical variables
    Standardize the magnitude of numeric features
    Split into testing and training datasets
    Apply scaler to the testing set
    
5.Modeling

    Fit Models with Training Data Set
    Review Model Outcomes — Iterate over additional models as needed.
    Identify the Final Model

6.Documentation

    Review the Results
    Present and share your findings - storytelling
    Finalize Code
    Finalize Documentation

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
print(os.getcwd())
df=pd.read_csv('data/step3_output.csv')
print(df.shape)
df.head(5)

C:\Users\SANJAY\Springboard\Guided-capstone-Unit-6
(73, 28)


Unnamed: 0  Unnamed: 0.1                        Name        state  \
0          15            15         Mt. Shasta Ski Park   California   
1          43            43    Sunlight Mountain Resort     Colorado   
2          49            49  Mount Southington Ski Area  Connecticut   
3          50            50           Powder Ridge Park  Connecticut   
4          51            51                 Ski Sundown  Connecticut   

   summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  ...  \
0         6890           1435      0        0.0          0          0  ...   
1         9895           2010      0        0.0          0          0  ...   
2          525            425      0        0.0          0          0  ...   
3          720            550      0        0.0          0          0  ...   
4         1075            625      0        0.0          0          0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0              425.0           225.0             140.0       34.0   
1              680.0            30.0             100.0       53.0   
2               51.0            51.0              63.0       55.0   
3               80.0            68.0              80.0       60.0   
4               70.0            70.0              84.0       50.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0            300.0          49.0          59.0              130.0   
1            250.0          65.0          65.0              135.0   
2             80.0          60.0          60.0               95.0   
3             80.0          45.0          55.0              100.0   
4             45.0          62.0          62.0               95.0   

   NightSkiing_ac  clusters  
0             0.0         1  
1             0.0         1  
2            51.0         0  
3            40.0         0  
4            66.0         0  

[5 rows x 28 columns]

In [3]:
#Create dummy variables for state. Add the dummies back to the dataframe and remove the original column for state.
df = pd.concat([df.drop(columns=['state']), pd.get_dummies(df['state'])], axis=1)
df.columns

Index(['Unnamed: 0', 'Unnamed: 0.1', 'Name', 'summit_elev', 'vertical_drop',
       'trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple',
       'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters',
       'California', 'Colorado', 'Connecticut', 'Idaho', 'Indiana', 'Iowa',
       'Maine', 'Massachusetts', 'Michigan', 'Minnesota', 'Missouri',
       'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania', 'Rhode Island',
       'South Dakota', 'Vermont', 'Virginia', 'Washington', 'West Virginia',
       'Wisconsin', 'Wyoming'],
      dtype='object')

## Standardize the magnitude of numeric features

Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set <b> AdultWeekend </b> to the y variable as our response for scaling and modeling. Later we will go back and consider the <b> AdultWeekday, dayOpenLastYear, and projectedDaysOpen </b>. For now leave them in the development dataframe.

In [4]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
x = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(x)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(x) 

## Split into training and testing datasets


Using sklearn model selection import train_test_split, and create a 75/25 split with the y = <b>AdultWeekend</b>. We will start by using the adult weekend ticket price as our response variable for modeling.

In [5]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train,X_test,y_train,y_test = train_test_split(X_scaled, y , test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.


# Step 5. Modeling

In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.

The Data Science Method¶

Problem Identification

Data Wrangling

Exploratory Data Analysis

Pre-processing and Training Data Development

Modeling
    
    Fit Models with Training Data Set
    Review Model Outcomes — Iterate over additional models as needed.
    Identify the Final Model

Documentation

    Review the Results
    Present and share your findings - storytelling
    Finalize Code
    Finalize Documentation

## Fit Models with a Training Dataset


Using sklearn, fit the model on your training dataset

<b>Model 1 </b>

In [6]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
model

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value.

Hint: you will have to use the predict() method 

In [7]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)
print("y_test: \n",y_test)
print("\ny_pred: \n",y_pred)

y_test: 
 [45.         33.         55.         41.         68.         47.
 67.         47.         35.         64.16681004 57.         39.
 60.         62.         50.         40.         65.         59.
 42.        ]

y_pred: 
 [45.41278959 46.99216377 41.12240726 46.65618209 39.62515661 57.08018467
 34.91841283 50.99952859 46.72297559 65.29766081 60.90011735 37.96070347
 58.59142289 59.37471317 56.97728877 43.4977057  51.21625218 54.48608282
 37.65969686]


## Review Model Outcomes — Iterate over additional models as needed

In [8]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

explained_variance_score(y_test, y_pred)

-0.17841467924139098

In [9]:
mean_absolute_error(y_test, y_pred)


8.600786253317091

Print the intercept value from the linear model.

Hint: our linear regression model lm has an attribute intercept_ for the intercept

In [10]:
print(model.intercept_)


51.6968692814076


The intercept is the mean <b>AdultWeekend</b> price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting <b>AdultWeekend</b> value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [11]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), x.columns, columns=['Coefficient'])

Coefficient
Unnamed: 0         4.812226e+13
Unnamed: 0.1       4.812226e+13
summit_elev        1.140234e+01
vertical_drop      4.570312e-01
trams              1.743164e-01
fastEight          6.054688e-02
fastSixes          1.435547e-01
fastQuads          8.779297e-01
quad               1.743652e+00
triple             1.796875e+00
double             3.125000e-02
surface            1.743530e+00
total_chairs       3.122559e-01
Runs               9.259766e+00
TerrainParks       1.727539e+00
LongestRun_mi      5.009766e-01
SkiableTerrain_ac  4.960938e-01
Snow Making_ac     6.562500e-01
daysOpenLastYear   6.524658e-01
yearsOpen          1.134644e-01
averageSnowfall    1.196289e-01
AdultWeekday       5.323730e+00
projectedDaysOpen  7.197266e-01
NightSkiing_ac     1.209961e+00
clusters           7.179688e+00
California         3.582031e+00
Colorado           2.447266e+00
Connecticut        1.083984e+00
Idaho              1.582031e-01
Indiana            1.687500e+00
Iowa               2.142578e+00
Maine              2.514648e+00
Massachusetts      1.176758e+00
Michigan           2.627930e+00
Minnesota          2.837891e+00
Missouri           3.125000e-02
New Hampshire      9.158936e-01
New Jersey         8.052979e-01
New Mexico         3.611084e+00
New York           3.050781e+00
North Carolina     7.172852e+00
Ohio               2.099121e+00
Oregon             3.373291e+00
Pennsylvania       4.843750e-01
Rhode Island       2.227369e-16
South Dakota       1.356445e+00
Vermont            1.862305e+00
Virginia           4.119141e+00
Washington         2.098633e+00
West Virginia      1.515625e+00
Wisconsin          1.888672e+00
Wyoming            3.312500e+00

### Model 2

You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results.

Hint: Try to construct another model using exactly the steps we followed above.

In [12]:
df = pd.read_csv("data/step3_output.csv")


In [13]:

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','state','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

In [14]:
# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [15]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X_train,y_train)

In [16]:
model2.score(X_test, y_test)

0.19451166006619025

In [17]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model2.predict(X_test)

In [18]:
explained_variance_score(y_test, y_pred)

0.19761659532526288

In [19]:
mean_absolute_error(y_test, y_pred)

7.470076646743068

In [20]:
pd.DataFrame(abs(lm2.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient')

Coefficient
fastSixes          0.000000e+00
fastEight          4.440892e-16
trams              2.664535e-15
projectedDaysOpen  4.507594e-02
triple             6.639386e-02
daysOpenLastYear   1.590309e-01
NightSkiing_ac     1.830775e-01
total_chairs       2.295693e-01
surface            4.003844e-01
fastQuads          4.847350e-01
Unnamed: 0.1       4.954341e-01
Unnamed: 0         4.954341e-01
yearsOpen          6.783328e-01
quad               7.217288e-01
double             9.802938e-01
LongestRun_mi      1.280287e+00
Snow Making_ac     1.445408e+00
Runs               1.566316e+00
averageSnowfall    1.658343e+00
summit_elev        1.675077e+00
clusters           2.051364e+00
SkiableTerrain_ac  2.231150e+00
TerrainParks       2.624564e+00
vertical_drop      3.465363e+00
AdultWeekday       7.222739e+00

When reviewing our new model coefficients, we see summit_elev is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with base_elev and vertical_drop. This time, rebuild the model without the state features and without the summit_elev and without base_elevand compare the results

## Model 3

In [21]:
df = pd.read_csv("data/step3_output.csv")


In [22]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','state','summit_elev','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

In [23]:

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [24]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X_train,y_train)

In [25]:
y_pred = model3.predict(X_test)


In [26]:
explained_variance_score(y_test, y_pred)


0.1526587682672247

In [27]:
mean_absolute_error(y_test, y_pred)


7.414692188987168

Identify the Final Model
Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone.

# Identify the final model

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1.| -0.17841467924139098 0. |8.600786253317091  |-|
| Model 2. |0.19761659532526288 0.|7.470076646743068 |'state'|
| Model 3. |0.1526587682672247 0. |7.414692188987168 |'state','summit_elev','base_elev'|